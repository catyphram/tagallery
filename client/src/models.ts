export enum LIST_MODE {
    MODE_VIEW,
    MODE_VERIFY,
}

export interface Category {
    name: string;
    key: string;
    description?: string;
}

export interface Image {
  file: string;
  assignedCategories?: string[];
  proposedCategories?: string[];
  starredCategory?: string;
}

export interface Config {
  api: string;
}
