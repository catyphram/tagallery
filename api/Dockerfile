FROM golang:1-alpine

RUN  apk add --update musl-dev gcc git

WORKDIR /go/src/app
COPY . .

RUN go get -d ./...
RUN go build

CMD ["./api"]