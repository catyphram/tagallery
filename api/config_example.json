{
    "database": "tagallery",
    "databaseHost": "localhost:27017",
    "port": 3333,
    "unprocessedImages": "/path/to/new/images"
    
}