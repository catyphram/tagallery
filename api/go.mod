module tagallery.com/api

go 1.15

require (
	github.com/gin-gonic/gin v1.6.3
	github.com/pkg/errors v0.9.1
	go.mongodb.org/mongo-driver v1.4.4
	go.uber.org/zap v1.16.0
)
