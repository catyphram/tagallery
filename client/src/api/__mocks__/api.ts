import { Category, Image } from '@/models';

export const loadCategories = jest.fn(async (): Promise<Category[]> => {
  return await [
    { name: 'Category 1', key: 'cat1', description: 'Category 1' },
    { name: 'Category 2', key: 'cat2', description: 'Category 2' },
    { name: 'Category 3', key: 'cat3', description: 'Category 3' },
    { name: 'Category 4', key: 'cat4', description: 'Category 4' },
    { name: 'Category 5', key: 'cat5', description: 'Category 5' },
    { name: 'Category 6', key: 'cat6', description: 'Category 6' },
  ];
});

export const loadImages = jest.fn(async (): Promise<Image[]> => {
  return await [
    { file: 'image0.jpg' },
    { file: 'image1.jpg' },
    { file: 'image2.jpg' },
    { file: 'image3.jpg' },
    { file: 'image4.jpg' },
    { file: 'image5.jpg' },
    { file: 'image6.jpg' },
    { file: 'image7.jpg' },
    { file: 'image8.jpg' },
    { file: 'image9.jpg' },
  ] as Image[];
});

export const updateImage = jest.fn(async (image): Promise<boolean> => {
  return await true;
});
