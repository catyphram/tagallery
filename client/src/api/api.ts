import { Category, Image } from '@/models';

/**
 * loadCategories loads the categories from the API.
 */
export const loadCategories = async (): Promise<Category[]> => {
  return await [
      { name: 'Category 1', key: 'cat1', description: 'Category 1' },
      { name: 'Category 2', key: 'cat2', description: 'Category 2' },
  ];
};

// Additional parameter for the local stub only so the URL differs on reload
// and Vue updates the images in the HTML.
let counter = 0;

/**
 * loadImages loads a batch of images from the API.
 * @todo: Add parameters for categories, mode and lastImage filter
 * Also don't forget to cancel the fetch request when another request is sent.
 */
export const loadImages = async (): Promise<Image[]> => {
  const stubImages = [] as Image[];
  for (let i = 0; i < 30; i++) {
    stubImages.push({
      file: `https://picsum.photos/400/400/?random&key=${counter}-${i}`,
    } as Image);
  }

  counter++;

  return await stubImages;
};
