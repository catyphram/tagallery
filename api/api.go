package main

import "tagallery.com/api/server"

func main() {
	server.StartServer()
}
