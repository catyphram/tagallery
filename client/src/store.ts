import Vue from 'vue';
import Vuex from 'vuex';

import * as api from '@/api/api';
import { Category, Image, LIST_MODE } from './models';

Vue.use(Vuex);

export const state = {
  categories: {
    loading: false,
    data: [] as Category[],
  },
  images: {
    loading: false,
    completed: false,
    data: [] as Image[],
  },
  listMode: LIST_MODE.MODE_VIEW,
  selectedCategories: [] as Category[],
  listUncategorized: false,
};

export const getters = {
  isCategorySelected: (state) => (category: Category) => state.selectedCategories.includes(category),
};

export const mutations = {
  updateCategories(state, { categories, loading = false }: { categories?: Category[], loading?: boolean }) {
    state.categories.data = categories;
    state.categories.loading = loading;
  },
  updateImages(state, {
    images, loading = false, completed = false, append = false,
  }: { images: Image[], loading?: boolean, completed?: boolean, append?: boolean }) {
    if (append) {
      state.images.data.push(...images);
    } else {
      state.images.data = images;
    }
    state.images.loading = loading;
    state.images.completed = completed;
  },
  setMode(state, { mode }: {mode: LIST_MODE}) {
    state.listMode = mode;
  },
  selectCategory(state, { category }: { category: Category }) {
    if (!state.selectedCategories.includes(category)) {
      state.selectedCategories.push(category);
      state.listUncategorized = false;
    }
  },
  unselectCategory(state, { category }: { category: Category }) {
    for (let i = 0; i < state.selectedCategories.length; i++) {
      if (state.selectedCategories[i] === category) {
        state.selectedCategories.splice(i, 1);
        break;
      }
    }
  },
  setListUncategorized(state, { flag }: { flag: boolean }) {
    state.listUncategorized = flag;
    if (flag) {
      state.selectedCategories = [];
    }
  },
};

export const actions = {
  async loadCategories(context) {
    context.commit('updateCategories', {
      categories: [],
      loading: true,
    });
    context.commit('updateCategories', {
      categories: await api.loadCategories(),
      loading: false,
    });
    context.dispatch('loadImages');
  },
  async loadImages(context, append = false) {
    context.commit('updateImages', {
      images: [],
      loading: true,
      append,
    });

    const images = await api.loadImages();
    context.commit('updateImages', {
      images,
      loading: false,
      completed: !images.length,
      append,
    });
  },
  setMode(context, mode: LIST_MODE) {
    context.commit('setMode', { mode });
    context.dispatch('loadImages');
  },
  toggleCategory(context, category: Category) {
    if (context.state.selectedCategories.includes(category)) {
      context.commit('unselectCategory', { category });
    } else {
      context.commit('selectCategory', { category });
    }
    context.dispatch('loadImages');
  },
  toggleListUncategorized(context) {
    context.commit('setListUncategorized', { flag: !context.state.listUncategorized });
    context.dispatch('loadImages');
  },
};

export default new Vuex.Store({
  state,
  getters,
  mutations,
  actions,
});
