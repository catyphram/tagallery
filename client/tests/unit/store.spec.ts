import { getters, mutations, actions, state as defaultState } from '@/store';
import { Category, LIST_MODE, Image } from '@/models';
import * as api from '@/api/api';

jest.mock('@/api/api');

describe('store.ts', () => {
  describe('getters', () => {
    it('isCategorySelected should return true if a category is found in selectedCategories', () => {
      const state = {
        ...defaultState,
        selectedCategories: [
          { name: 'Category 1' }, { name: 'Category 2' }, { name: 'Category 3' },
        ] as Category[],
      };
      expect(getters.isCategorySelected(state)(state.selectedCategories[0])).toBe(true);
    });
  });

  describe('mutations', () => {
    it('updateCategories should set the categories and loading status in the state', () => {
      const categories = [{ name: 'Category 1' }, { name: 'Category 2' } ] as Category[];
      const state = {
        ...defaultState,
        categories: {
          data: [] as Category[],
          loading: true,
          error: undefined,
        },
      };
      mutations.updateCategories(state, {
        categories,
      });

      expect(state.categories.data).toBe(categories);
      expect(state.categories.loading).toBe(false);
      expect(state.categories.error).toBeUndefined();

      mutations.updateCategories(state, {
        categories: [],
        error: 'error',
      });

      expect(state.categories.error).toBe('error');
    });

    it('updateImages should set the images, completed and loading status in the state', () => {
      const images = [{ file: 'test1.jpg' }, { file: 'test2.jpg' } ] as Image[];
      const state = {
        ...defaultState,
        images: {
          data: [] as Image[],
          loading: true,
          completed: false,
        },
      };
      mutations.updateImages(state, {
        images,
        loading: false,
        completed: true,
      });

      expect(state.images.data).toBe(images);
      expect(state.images.loading).toBe(false);
      expect(state.images.completed).toBe(true);
    });

    it('updateImages should append the images if #append is true', () => {
      const images = [{ file: 'test1.jpg' }, { file: 'test2.jpg' } ] as Image[];
      const newImages = [{ file: 'test3.jpg' }, { file: 'test4.jpg' } ] as Image[];
      const state = {
        ...defaultState,
        images: {
          completed: false,
          data: [...images],
          loading: true,
        },
      };
      mutations.updateImages(state, {
        images: newImages,
        append: true,
      });

      expect(state.images.data).toEqual(images.concat(newImages));
    });

    it('selectImage should select the image', () => {
      const state = {
        ...defaultState,
      };
      mutations.selectImage(state, { index: 1 });

      expect(state.selectedImage).toBe(1);
    });

    it('setMode should set the mode in the state', () => {
      const state = {
        ...defaultState,
        listMode: LIST_MODE.MODE_VIEW,
      };
      mutations.setMode(state, { mode: LIST_MODE.MODE_VERIFY });

      expect(state.listMode).toBe(LIST_MODE.MODE_VERIFY);
    });

    it('selectCategory should add the category in the state and unset #listUncategorized', () => {
      const category = { name: 'Category 1' } as Category;
      const state = {
        ...defaultState,
        selectedCategories: [] as Category[],
        listUncategorized: true,
      };

      mutations.selectCategory(state, { category });
      expect(state.selectedCategories).toEqual([category ]);
      expect(state.listUncategorized).toBe(false);

      mutations.selectCategory(state, { category });
      expect(state.selectedCategories).toEqual([category ]);
    });

    it('unselectCategory should remove a category if it is selected', () => {
      const cat1 = { name: 'Category 1' } as Category;
      const cat2 = { name: 'Category 2' } as Category;
      const cat3 = { name: 'Category 3' } as Category;
      const categories = [cat1, cat2, cat3 ];
      const state = {
        ...defaultState,
        selectedCategories: categories,
      };

      mutations.unselectCategory(state, { category: { name: 'Category 0' } as Category });
      expect(state.selectedCategories).toEqual([cat1, cat2, cat3 ]);

      mutations.unselectCategory(state, { category: cat2 });
      expect(state.selectedCategories).toEqual([cat1, cat3 ]);
    });

    it('setListUncategorized should change the flag in the state and remove selected categories if true', () => {
      const state = {
        selectedCategories: [
          { name: 'Category 1' }, { name: 'Category 2' }, { name: 'Category 3' },
        ] as Category[],
        listUncategorized: false,
      };

      mutations.setListUncategorized(state, { flag: true });
      expect(state.selectedCategories).toEqual([]);
      expect(state.listUncategorized).toBe(true);

      mutations.setListUncategorized(state, { flag: false });
      expect(state.listUncategorized).toBe(false);
    });
  });

  describe('actions', () => {
    let context;

    beforeEach(() => {
      context = {
        commit: jest.fn(),
        dispatch: jest.fn(),
      };
    });

    it('loadCategories should load the categories', async () => {
      await actions.loadCategories(context);
      expect(context.commit).toMatchSnapshot();
      (api.loadCategories as jest.Mock).mockImplementationOnce(() => {
        return Promise.reject('error');
      });
      await actions.loadCategories(context);
      expect(context.commit).toMatchSnapshot();

    });

    it('loadImages should load a batch of images', async () => {
      await actions.loadImages(context, true);
      expect(context.commit).toMatchSnapshot();
      await actions.loadImages(context);
      expect(context.commit).toMatchSnapshot();
    });

    it('setMode should set the viewing mode', () => {
      actions.setMode(context, LIST_MODE.MODE_VERIFY);
      expect(context.commit).toHaveBeenCalledWith('setMode', { mode: LIST_MODE.MODE_VERIFY });
    });

    it('toggleCategory should select a category if it is not selected yet', () => {
      const category = { name: 'Category 1', key: 'cat1' } as Category;
      context.state = {
        selectedCategories: [],
      };
      actions.toggleCategory(context, category);
      expect(context.commit).toHaveBeenCalledWith('selectCategory', { category });
    });

    it('toggleCategory should unselect a category if it is selected already', () => {
      const category = { name: 'Category 1', key: 'cat1' } as Category;
      context.state = {
        selectedCategories: [category ],
      };
      actions.toggleCategory(context, category);
      expect(context.commit).toHaveBeenCalledWith('unselectCategory', { category });
    });

    it('toggleListUncategorized should toggle the flag', () => {
      context.state = {
        listUncategorized: false,
      };
      actions.toggleListUncategorized(context);
      expect(context.commit).toHaveBeenCalledWith('setListUncategorized', { flag: true });
    });
  });
});
