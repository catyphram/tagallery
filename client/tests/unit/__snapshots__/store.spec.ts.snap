// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store.ts actions loadCategories should load the categories 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "updateCategories",
      Object {
        "categories": Array [],
        "loading": true,
      },
    ],
    Array [
      "updateCategories",
      Object {
        "categories": Array [
          Object {
            "description": "Category 1",
            "key": "cat1",
            "name": "Category 1",
          },
          Object {
            "description": "Category 2",
            "key": "cat2",
            "name": "Category 2",
          },
          Object {
            "description": "Category 3",
            "key": "cat3",
            "name": "Category 3",
          },
          Object {
            "description": "Category 4",
            "key": "cat4",
            "name": "Category 4",
          },
          Object {
            "description": "Category 5",
            "key": "cat5",
            "name": "Category 5",
          },
          Object {
            "description": "Category 6",
            "key": "cat6",
            "name": "Category 6",
          },
        ],
        "loading": false,
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`store.ts actions loadImages should load a batch of images 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "updateImages",
      Object {
        "append": true,
        "images": Array [],
        "loading": true,
      },
    ],
    Array [
      "updateImages",
      Object {
        "append": true,
        "completed": false,
        "images": Array [
          Object {
            "file": "image0.jpg",
          },
          Object {
            "file": "image1.jpg",
          },
          Object {
            "file": "image2.jpg",
          },
          Object {
            "file": "image3.jpg",
          },
          Object {
            "file": "image4.jpg",
          },
          Object {
            "file": "image5.jpg",
          },
          Object {
            "file": "image6.jpg",
          },
          Object {
            "file": "image7.jpg",
          },
          Object {
            "file": "image8.jpg",
          },
          Object {
            "file": "image9.jpg",
          },
        ],
        "loading": false,
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`store.ts actions loadImages should load a batch of images 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "updateImages",
      Object {
        "append": true,
        "images": Array [],
        "loading": true,
      },
    ],
    Array [
      "updateImages",
      Object {
        "append": true,
        "completed": false,
        "images": Array [
          Object {
            "file": "image0.jpg",
          },
          Object {
            "file": "image1.jpg",
          },
          Object {
            "file": "image2.jpg",
          },
          Object {
            "file": "image3.jpg",
          },
          Object {
            "file": "image4.jpg",
          },
          Object {
            "file": "image5.jpg",
          },
          Object {
            "file": "image6.jpg",
          },
          Object {
            "file": "image7.jpg",
          },
          Object {
            "file": "image8.jpg",
          },
          Object {
            "file": "image9.jpg",
          },
        ],
        "loading": false,
      },
    ],
    Array [
      "updateImages",
      Object {
        "append": false,
        "images": Array [],
        "loading": true,
      },
    ],
    Array [
      "updateImages",
      Object {
        "append": false,
        "completed": false,
        "images": Array [
          Object {
            "file": "image0.jpg",
          },
          Object {
            "file": "image1.jpg",
          },
          Object {
            "file": "image2.jpg",
          },
          Object {
            "file": "image3.jpg",
          },
          Object {
            "file": "image4.jpg",
          },
          Object {
            "file": "image5.jpg",
          },
          Object {
            "file": "image6.jpg",
          },
          Object {
            "file": "image7.jpg",
          },
          Object {
            "file": "image8.jpg",
          },
          Object {
            "file": "image9.jpg",
          },
        ],
        "loading": false,
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
